### ISSUE TITLE
RNN Benchmark derived from rnn tutorial

### ISSUE THREAD

**Issue by David-Levinthal** [tensorflow/models#1983]
------------------------

### System information
- **What is the top-level directory of the model you are using**:   RNN
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:from source
- **TensorFlow version (use command below)**:R1.2
- **Bazel version (if compiling from source)**:0.4.5
- **CUDA/cuDNN version**:8.0/6.0
- **GPU model and memory**:P4
- **Exact command to reproduce**:

### Describe the problem
We are trying to identify components that can be used to construct a HW evaluation performance benchmark for machine learning. For the Tensorflow component we are using the tf_cnn_benchmarks and while looking for an rnn component identified the tutorials for rnns (PTB and Translate) as likely candidates for coverage of LSTM and GRU rnn's. There are issues with translate
[tensorflow/models] wrong with `tutorials/rnn/translate/seq2seq_model.py` (#1875)
Toby Boyd suggested creating a new ticket to track progress of the RNN benchmark development
He wrote:
On Jul 11, 2017 6:04 PM, "Toby Boyd" <notifications@github.com> wrote:
The perf team should be updating ptb soon, if my memory is correct. We
have some tweaks to have it use cuDNN, which should make it a much better
measure. If you want to track the progress or track getting a good RNN
benchmark, open a ticket and assign to me or mention me and I will assign
it to myself and update it with progress. Glad/hope you find the benchmark
tool useful.

Toby

so this issue was created

**Comment by tfboyd** [comment]
Assigned to me.  thank you.  This is in progress, I hope I can get it pushed out a few weeks.

**Comment by David-Levinthal** [comment]
that would be great
please let me know how I can assist.
Creating a solid HW benchmark suite that can evaluate throughput, latency
and convergence speed as a function of precision is sorely needed
d

**Comment by David-Levinthal** [comment]
short question. In ptb_word_lm.py, if I change:
#        valid_perplexity = run_epoch(session, mvalid)
to
        valid_perplexity = run_epoch(session, mvalid, eval_op=None,
                                     verbose=True)
the invocation appears to produces the verbose output for the validation and I believe the WPS rate for validation would be equal to the value for inference (at the same precision) would it not?
Epoch: 1 Learning rate: 1.000
0.004 perplexity: 6113.837 speed: 3718 wps
0.104 perplexity: 838.424 speed: 8786 wps
0.204 perplexity: 620.519 speed: 9077 wps
0.304 perplexity: 500.676 speed: 9184 wps
0.404 perplexity: 432.075 speed: 9237 wps
0.504 perplexity: 386.909 speed: 9274 wps
0.604 perplexity: 348.295 speed: 9298 wps
0.703 perplexity: 321.807 speed: 9320 wps
0.803 perplexity: 300.869 speed: 9332 wps
0.903 perplexity: 281.499 speed: 9344 wps
Epoch: 1 Train Perplexity: 267.122
0.054 perplexity: 171.932 speed: 15779 wps
0.152 perplexity: 176.262 speed: 18708 wps
0.250 perplexity: 183.543 speed: 19663 wps
0.348 perplexity: 182.416 speed: 19967 wps
0.446 perplexity: 178.416 speed: 20181 wps
0.543 perplexity: 177.023 speed: 20225 wps
0.641 perplexity: 177.986 speed: 20186 wps
0.739 perplexity: 177.981 speed: 20274 wps
0.837 perplexity: 175.735 speed: 20295 wps
0.935 perplexity: 175.905 speed: 20364 wps
Epoch: 1 Valid Perplexity: 176.152

**Comment by David-Levinthal** [comment]
on a related note: it is far from obvious to me as to how I control the
OpenMP thread counts through
inter_op_parallelism_threads=FLAGS.inter_op,intra_op_parallelism_threads=FLAGS.intra_op
in ptb_word_lm.py
I have been stunningly unsuccessful.
if one uses an MKL build of TF not getting this right causes 2-10X
performance degradations
see:
https://stackoverflow.com/questions/45243543/how-to-set-inter-op-parallelism-threads-and-intra-op-parallelism-threads-for-see

??
thanks
d


On Tue, Jul 18, 2017 at 2:05 PM, David Levinthal <david.levinthal1@gmail.com

**Comment by David-Levinthal** [comment]
figured it out finally

the solution for ptb_word_lm.py seems to be to add

config =
tf.ConfigProto(inter_op_parallelism_threads=FLAGS.inter_op,intra_op_parallelism_threads=FLAGS.intra_op)

sess = tf.Session(config=config)

before the call to the ptb_reader thereby setting the session env. (along
with some OMP* env variables)

**Comment by tfboyd** [comment]
Hi David,

Sorry I did not see your email earlier.  Did you get a speedup using MKL on
ptb model?  I have not made time to test RNNs with MKL and am curious if
you had any good results.  Even if they are back of the napkin speedup
guesses I am very interested.

Toby

**Comment by David-Levinthal** [comment]
Toby
is there a place I can send a CSV ?
I suspect I should not attach it to the github bug
:-)
as it might be a bit too public

The speed ups are substantial..I am setting intra_op to the physical core
count and inter_op to 2
running with HT enabled

yeah I got quite confused

BTW..would it make any sense to replace
tf.contrib.rnn.BasicLSTMCell
with either
tf.nn.rnn_cell.LSTMCell
or
tf.nn.rnn_cell.GRUCell
in ptb_word_lm.py

per discussion in
https://r2rt.com/recurrent-neural-networks-in-tensorflow-ii.html
??




d

**Comment by tfboyd** [comment]
I think someone is looking at making that LSTMCell method change in the
example. There were a number of performance tweaks to the ptb example and
RNN examples in general we need to do.  There have been improvements since
the examples were made.  I am not sure if the change you are suggesting is
performance but I wanted to share what I know.  I only skimmed the link you
provided.  If you have time to do a PR to change the method being used that
would be great.  If not I understand and I hope we get to it soon.  There
might be some other tweaks.

If the data is not private to you feel free to just upload it to a github
repo or just paste it in text form unless you have a crazy amount of data.
You could also put it in a google document and share read-only publicly.
Public is totally fine for me.

**Comment by David-Levinthal** [comment]
Toby
I have no problem sending my modified version of ptb_word_lm.py for
OMP/MKL, the changes are minimal
but result in the script being specific to an MKL/OMP run
I am seeing about a 50% improvement over the default TF build with Eigen
numbers from the wps printout which I modified slightly to get a
verification wps
and the rounded

training  batch size  eigen R1.2  MKL R1.2
small           20          8400      13200
medium       20           2715       5000
large            20                        1750

verification
small           20          22900     35200
medium       20          11600      15800
large           20                           6500
excuse the formatting

**Comment by karmel** [comment]
@tfboyd / @David-Levinthal -- what is the status of this issue? There has been a lot of work in the last ~9 months since this was opened; is this thread still relevant?

**Comment by tfboyd** [comment]
I am going to close.  Time got away.  An RNN for MKL would be nice maybe more so as an inference example.  Even if I had the code now I do not have enough open time to merge it together into something coherent or benchmark it.  :-(  I do not like when this stuff slips through July-2017.  wow.
